Dedede Down B Rework and other adjustments

SirTanknSpank opened the pull request:
A new down b and some other big adjustments for the plus-sized penguin.

Assets: [dedede-assets.zip](https://github.com/user-attachments/files/16740137/dedede-assets.zip)



# Changelog


### Forward Tilt
The goal here is to make the move more consistent, rather than having people fall out before the final hit. 
- Multihits:
  - [/] Hammer hitboxes adjusted to be on the hammer2 bone, as opposed to the top bone
  - [/] Number of hitboxes: 5 -> 3
- Finisher:
  - [+] Hammer hitbox size: 7.0 units -> 7.5 units

Old:
![OldMultiFtilt](https://github.com/user-attachments/assets/084be37a-ef4e-4553-b98f-18b94f2b573e)
![OldFinalFtilt](https://github.com/user-attachments/assets/e4fa8cb1-7d43-4aa7-858a-08ac0bff2917)


New:

![NewFtilt](https://github.com/user-attachments/assets/70030417-4c80-4dda-aa8e-76440009318b)
![NewFinalFtilt](https://github.com/user-attachments/assets/4218ad98-db7a-4120-89e4-7e472a642054)

### Down Tilt
This should make it match P+ frame data
- [+] Animation speed increased
- [/] Active Frames: 7-20 -> 6-14
- [+] FAF: 37 -> 31

### Down Throw
- [-] BKB: 50 -> 65
- [/] KBG: 80 -> 60

### Up Throw
- [-] KBG: 80 -> 90

### Up Special
- [!] Now stops all momentum before the jump
- [!] Can now sweetspot ledge before the jump
- [+] Landing star offset: 7 units -> 9 units

### Side Special
A small buff, as well as making smash gordo have more of a difference compared to tilt gordo.
Also a QoL change to make it so aerial down angled gordo no longer forces a fastfall. 
- [+] Smash Gordo Speed Multiplier: 1.25x -> 1.6x
- [+] FAF: 55 -> 51
- [/] Fastfall window: Anytime lmao -> Frame 19+

### Gordo
Partial revert of gordo to have knockback scaling again. This should make it better at higher %'s while still keeping it low knockback all around
- [+] Shieldstun Multiplier: 0.7x -> 0.8x
- [/] FKB: 60 -> 0
- [/] KBG: 110 -> 55
- [/] BKB: 0 -> 50

### Gordo Dash
These changes should make gordo dash feel *way* better and actually worth using
- [+] Air Dash Speed Multiplier: 1.5x -> 1.8x
- [+] Grounded dash speed multiplier: 1x -> 1.2x
- [+] FAF: 56 -> 42
- [/] Retoss Angle: 120 -> 60

### Down Special
Reworked into Jet Hammer Spin! This is a high risk move where Dedede takes out his jet hammer and launches himself forward while spinning. The move can be charged to deal more damage, as well as gaining an extra set of spins for every 30 frames of charge.



https://github.com/user-attachments/assets/fba0fd99-2364-4860-82bb-e1af6f63fd8f




https://github.com/user-attachments/assets/6d90ce96-bd4c-4fbc-acbc-e0133953bc3e





The move can be charged for 80 frames, giving a total of 3 sets of spins.


- Attack Data
  - Startup: 18 Frames
  - Damage: 14% , Maximum of 18% if fully charged
  - Angle: 361
  - KBG: 60
  - BKB: 85
  - Self Damage: 12% if fully charged
  - Goes into special fall if used in the air

Other important things to note:
- Armor is removed

### Side Taunt (Facing Right)
- [+++] Dedede will now equip his mask if special is held on frame 3

https://github.com/user-attachments/assets/62b309b4-2751-4bb7-b18f-480f6c1199d1

SirTanknSpank committed:
Init

SirTanknSpank committed:
Merge branch 'HDR-Development:dev' into fat-penguin-down-b-rework

SirTanknSpank committed:
Revive this from the dead

SirTanknSpank committed:
Bonk, multihit removal, flag rename and more

SirTanknSpank committed:
Merge branch 'HDR-Development:dev' into fat-penguin-down-b-rework

SirTanknSpank committed:
Effect and speed fixes

SirTanknSpank committed:
Merge branch 'HDR-Development:dev' into fat-penguin-down-b-rework

SirTanknSpank committed:
Actual mask toggle

and also removing an old comment

SirTanknSpank committed:
Smol changes

SirTanknSpank committed:
Merge branch 'fat-penguin-down-b-rework' of https://github.com/SirTanknSpank/HewDraw-Remix into fat-penguin-down-b-rework

SirTanknSpank committed:
Moving fastfall gordo to frame 19+

SirTanknSpank committed:
Whoops

SirTanknSpank committed:
Merge branch 'HDR-Development:dev' into fat-penguin-down-b-rework

WuBoytH commented on fighters/dedede/src/status/special_lw.rs:
```diff
@@ -9,6 +9,88 @@ unsafe extern "C" fn special_lw_jump_squat_exec(fighter: &mut L2CFighterCommon)
 }
 
+unsafe extern "C" fn special_lw_wait_pre(fighter: &mut L2CFighterCommon) -> L2CValue{
+    StatusModule::change_status_force(fighter.module_accessor, *FIGHTER_DEDEDE_STATUS_KIND_SPECIAL_LW_ATTACK, true);
```
instead of using change_status_force, please do the following:
```rs
unsafe extern "C" fn special_lw_wait_pre(fighter: &mut L2CFighterCommon) -> L2CValue{
    StatusModule::set_status_kind_interrupt(fighter.module_accessor, *FIGHTER_DEDEDE_STATUS_KIND_SPECIAL_LW_ATTACK);
    1.into()
}
```

This will skip running the pre status entirely and go straight to the ATTACK status.

WuBoytH commented on fighters/dedede/src/status/special_lw.rs:
```diff
@@ -9,6 +9,88 @@ unsafe extern "C" fn special_lw_jump_squat_exec(fighter: &mut L2CFighterCommon)
+
+    //Bonk upon hitting a shield
+    if AttackModule::is_infliction_status(fighter.module_accessor, *COLLISION_KIND_MASK_SHIELD){
+        fighter.change_status_req(*FIGHTER_DEDEDE_STATUS_KIND_SPECIAL_HI_FAILURE, false);
```
should be `fighter.change_status` when you're in a status script.

Also should have a return so that it doesn't run the rest of the exec status after the change status call.

WuBoytH commented on fighters/dedede/src/status/special_lw.rs:
```diff
@@ -9,6 +9,88 @@ unsafe extern "C" fn special_lw_jump_squat_exec(fighter: &mut L2CFighterCommon)
+        if StatusModule::situation_kind(fighter.module_accessor) == *SITUATION_KIND_AIR {
+            MotionModule::change_motion(fighter.module_accessor, Hash40::new("fall_special"), 0.0, 1.0, false, 0.0, false, false);
+
+            fighter.change_status(FIGHTER_STATUS_KIND_FALL_SPECIAL.into(), false.into()); 
```
same note for return.

WuBoytH commented on fighters/dedede/src/status/special_lw.rs:
```diff
@@ -9,6 +9,88 @@ unsafe extern "C" fn special_lw_jump_squat_exec(fighter: &mut L2CFighterCommon)
+
+    if MotionModule::is_end(fighter.module_accessor) &&
+    StatusModule::situation_kind(fighter.module_accessor) == SITUATION_KIND_AIR{
+        MotionModule::change_motion(fighter.module_accessor, Hash40::new("fall_special"), 0.0, 1.0, false, 0.0, false, false);
```
this line shouldn't be necessary if you're already changing into the FALL_SPECIAL status.

WuBoytH commented on fighters/dedede/src/status/special_hi.rs:
```diff
@@ -3,10 +3,15 @@ use super::*;
 // FIGHTER_DEDEDE_STATUS_KIND_SPECIAL_HI_FAILURE
 
 unsafe extern "C" fn special_hi_failure_main(fighter: &mut L2CFighterCommon) -> L2CValue{
```
this main status should turn into a pre one that looks similar to this
```rs
unsafe extern "C" fn special_hi_failure_pre(fighter: &mut L2CFighterCommon) -> L2CValue{
    if !VarModule::is_flag(fighter.battle_object, vars::dedede::instance::JET_GROUND_BONK) {
        StatusModule::set_status_kind_interrupt(fighter.module_accessor, *FIGHTER_STATUS_KIND_FALL_SPECIAL);
        return 1.into();
    }
    smashline::original_status(Pre, fighter, *FIGHTER_DEDEDE_STATUS_KIND_SPECIAL_HI_FAILURE)(fighter)
}
```

See below notes as to why.

SirTanknSpank committed:
Status Implementation Changes

WuBoytH commented on fighters/dedede/src/status/special_lw.rs:
```diff
@@ -9,6 +9,78 @@ unsafe extern "C" fn special_lw_jump_squat_exec(fighter: &mut L2CFighterCommon)
+
+    if MotionModule::frame(fighter.module_accessor) > 30.0 && StatusModule::is_situation_changed(fighter.module_accessor){
+        if StatusModule::situation_kind(fighter.module_accessor) == *SITUATION_KIND_AIR {
+            MotionModule::change_motion(fighter.module_accessor, Hash40::new("fall_special"), 0.0, 1.0, false, 0.0, false, false);
```
This line can be removed

WuBoytH commented on fighters/dedede/src/status/special_lw.rs:
```diff
@@ -9,6 +9,78 @@ unsafe extern "C" fn special_lw_jump_squat_exec(fighter: &mut L2CFighterCommon)
+unsafe extern "C" fn special_lw_attack_exec(fighter: &mut L2CFighterCommon) -> L2CValue{
+    //Continue spinning if the button is held down
+    if ControlModule::check_button_on(fighter.module_accessor, *CONTROL_PAD_BUTTON_SPECIAL) 
+    && MotionModule::frame(fighter.module_accessor) < 25.0 
```
Please make these frame checks a status flag that's enabled in ACMD instead

SirTanknSpank committed:
A ton of changes, it's basically a second rework

- Remove shield bonk
- Re-implement charging
- Reduce FAF By 3 frames
- Startup increased by 3 frames
- Re-implement status scripts
- Removing some unused variables

SirTanknSpank committed:
Changes for the new animations

SuddyN commented on fighters/dedede/src/acmd/specials.rs:
```diff
@@ -110,7 +110,7 @@ unsafe extern "C" fn game_specialsstart(agent: &mut L2CAgentBase) {
     }
     frame(lua_state, 25.0);
-    FT_MOTION_RATE(agent, 35.0 / (65.0-26.0));    
+    FT_MOTION_RATE(agent, 31.0 / (65.0-26.0));    
```
`FT_MOTION_RATE_RANGE(agent, 26.0, 65.0, 31.0)`

new standard is to use this function, rather than math. usage is `agent`, `start_frame`, `end_frame`, `duration`

SuddyN commented on fighters/dedede/src/acmd/specials.rs:
```diff
@@ -198,7 +198,7 @@ unsafe extern "C" fn game_specialairsstart(agent: &mut L2CAgentBase) {
     }
     frame(lua_state, 25.0);
-    FT_MOTION_RATE(agent, 35.0 / (65.0-26.0));    
+    FT_MOTION_RATE(agent, 31.0 / (65.0-26.0));    
```
`FT_MOTION_RATE_RANGE(agent, 26.0, 65.0, 31.0)`

SuddyN commented on fighters/dedede/src/acmd/specials.rs:
```diff
@@ -336,8 +336,69 @@ unsafe extern "C" fn game_speciallwstart(agent: &mut L2CAgentBase) {
+        }
+    }
+    frame(lua_state, 9.0);
+    FT_MOTION_RATE(agent, 7.0 / (27.0-10.0));
```
`FT_MOTION_RATE_RANGE(agent, 10.0, 27.0, 7.0)`

SuddyN commented on fighters/dedede/src/acmd/specials.rs:
```diff
@@ -346,42 +407,188 @@ unsafe extern "C" fn game_specialairlwstart(agent: &mut L2CAgentBase) {
+        }
+    }
+    frame(lua_state, 9.0);
+    FT_MOTION_RATE(agent, 7.0 / (27.0-10.0));
```
`FT_MOTION_RATE_RANGE(agent, 10.0, 27.0, 7.0)`

SuddyN commented on fighters/dedede/src/acmd/specials.rs:
```diff
@@ -346,42 +407,188 @@ unsafe extern "C" fn game_specialairlwstart(agent: &mut L2CAgentBase) {
+    let charge_frames = VarModule::get_int(agent.battle_object, vars::dedede::instance::JET_CHARGE_FRAMES);
+    let charge_level = charge_frames as f32 / 30.0;
+
+    if is_excute(agent) {
+        FT_MOTION_RATE(agent, 45.0/(23.0));
+    }
```
avoid using `FT_MOTION_RATE` or similar motion-rating functions *inside* an `if is_excute(agent)` conditional. This often causes problems for statuses that transition between air/ground motions when landing/falling, because the conditional fails in those circumstances

rewrite by just removing the conditional and keeping the motion rate

SuddyN commented on fighters/dedede/src/acmd/specials.rs:
```diff
@@ -346,42 +407,188 @@ unsafe extern "C" fn game_specialairlwstart(agent: &mut L2CAgentBase) {
+    }
+    frame(lua_state, 19.0);
     if is_excute(agent) {
-        FT_MOTION_RATE(agent, 43.0/(60.0-15.0));
+        MotionModule::set_rate(agent.module_accessor, 1.0);
     }
```
avoid using `FT_MOTION_RATE` or similar motion-rating functions inside an `if is_excute(agent)` conditional. this applies to `MotionModule::set_rate` too!

SuddyN commented on fighters/dedede/src/acmd/specials.rs:
```diff
@@ -346,42 +407,188 @@ unsafe extern "C" fn game_specialairlwstart(agent: &mut L2CAgentBase) {
+            ATTACK(agent, 0, 0, Hash40::new("hammer2"), 11.0 + (charge_level * 1.5), 361, 60, 0, 85, 8.0, 0.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_F, false, 10, 0.0, 0.0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL ...
+            ATTACK(agent, 1, 0, Hash40::new("hammer1"), 11.0 + (charge_level * 1.5), 361, 60, 0, 85, 4.0, 10.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_F, false, 10, 0.0, 0.0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVE ...
+            ATTACK(agent, 2, 0, Hash40::new("hammer1"), 11.0 + (charge_level * 1.5), 361, 60, 0, 85, 4.0, 5.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_F, false, 10, 0.0, 0.0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL ...
+            ATTACK(agent, 3, 0, Hash40::new("hammer1"), 11.0 + (charge_level * 1.5), 361, 60, 0, 85, 4.0, 0.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_F, false, 10, 0.0, 0.0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL ...
+        }
+        else{
+            ATTACK(agent, 0, 0, Hash40::new("hammer2"), 14.0 + (charge_level * 1.5), 361, 60, 0, 85, 8.0, 0.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_F, false, 10, 0.0, 0.0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL ...
+            ATTACK(agent, 1, 0, Hash40::new("hammer1"), 14.0 + (charge_level * 1.5), 361, 60, 0, 85, 4.0, 10.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_F, false, 10, 0.0, 0.0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVE ...
+            ATTACK(agent, 2, 0, Hash40::new("hammer1"), 14.0 + (charge_level * 1.5), 361, 60, 0, 85, 4.0, 5.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_F, false, 10, 0.0, 0.0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL ...
+            ATTACK(agent, 3, 0, Hash40::new("hammer1"), 14.0 + (charge_level * 1.5), 361, 60, 0, 85, 4.0, 0.0, 0.0, 0.0, None, None, None, 1.0, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_F, false, 10, 0.0, 0.0, false, false, false, false, true, *COLLISION_SITUATION_MASK_GA, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_normal"), *ATTACK_SOUND_LEVEL ...
+        }
+        
     }
-    frame(lua_state, 13.0);
+    frame(lua_state, 20.0);
     if is_excute(agent) {
-        AttackModule::clear_all(boma);
+        if !VarModule::is_flag(agent.battle_object, vars::dedede::instance::CONTINUE_JET_SPIN) && !agent.is_situation(*SITUATION_KIND_AIR){
+            KineticModule::add_speed(agent.module_accessor, &Vector3f{x: 0.6, y:0.0, z:0.0});
+        }
     }
-    frame(lua_state, 15.0);
+    frame(lua_state, 18.0);
+    if is_excute(agent){
+        VarModule::set_flag(agent.battle_object, vars::dedede::status::SPECIAL_LW_CONTINUE_SPIN, true);
+    }
+    frame(lua_state, 19.0);
     if is_excute(agent) {
-        FT_MOTION_RATE(agent, 43.0/(60.0-15.0));
+        MotionModule::set_rate(agent.module_accessor, 1.0);
     }
-    frame(lua_state, 60.0);
+    frame(lua_state, 23.0);
     if is_excute(agent) {
-        FT_MOTION_RATE(agent, 1.0);
+        KineticModule::clear_speed_all(boma);
+        KineticModule::enable_energy(agent.module_accessor, *FIGHTER_KINETIC_ENERGY_ID_CONTROL);
+        AttackModule::clear_all(agent.module_accessor);
+        FT_MOTION_RATE(agent, 7.0/(31.0 - 22.0));
+    }
```
bump. motion rating is done using the math method and also inside a conditional here.

SuddyN commented on fighters/dedede/src/acmd/specials.rs:
```diff
@@ -415,31 +622,113 @@ unsafe extern "C" fn game_speciallwmax(agent: &mut L2CAgentBase) {
+    let charge_frames = VarModule::get_int(agent.battle_object, vars::dedede::instance::JET_CHARGE_FRAMES);
+    let charge_level = charge_frames as f32 / 30.0;
+
+    if is_excute(agent) {
+        FT_MOTION_RATE(agent, 45.0/(23.0));
+        KineticModule::clear_speed_all(boma);
+    }
```
bump

SuddyN commented on fighters/dedede/src/acmd/specials.rs:
```diff
@@ -415,31 +622,113 @@ unsafe extern "C" fn game_speciallwmax(agent: &mut L2CAgentBase) {
     }
-    frame(lua_state, 13.0);
+    frame(lua_state, 23.0);
     if is_excute(agent) {
+        KineticModule::clear_speed_all(boma);
+        FT_MOTION_RATE(agent, 7.0/(31.0 - 22.0));
         AttackModule::clear_all(boma);
     }
```
bump...

SuddyN commented on fighters/dedede/src/jethammer/acmd.rs:
```diff
@@ -0,0 +1,21 @@
+    let lua_state = agent.lua_state_agent;
+    let boma = agent.boma();
+
+    if is_excute(agent) {
+        FT_MOTION_RATE(agent, 2.0);
+    }
```
bump!

SuddyN commented on fighters/dedede/src/opff.rs:
```diff
@@ -119,6 +123,9 @@ unsafe fn gordo_recatch(boma: &mut BattleObjectModuleAccessor, frame: f32, fight
             if fighter.status_frame() > 1 && fighter.status_frame() < 4{
                 ControlModule::reset_main_stick(boma);
             }
+            if StatusModule::is_situation_changed(fighter.module_accessor){
+                MotionModule::set_rate(fighter.module_accessor, 1.4);
+            }
```
what are we doing here..? this will presumably just be overwritten by any motion rating done in the ACMD. maybe you are trying to solve a problem that is caused by placing motion rating calls inside an `if is_excute(agent)` conditional? Whatever it is, you probably do not need or want to set motion rate to 1.4 on `StatusModule::is_situation_changed`.

SuddyN commented on fighters/dedede/src/opff.rs:
```diff
@@ -129,12 +136,124 @@ unsafe fn gordo_recatch(boma: &mut BattleObjectModuleAccessor, frame: f32, fight
+                }
+            }
+
+            ModelModule::set_joint_translate(fighter.module_accessor, Hash40::new("shoulderr"), &Vector3f{x:3.0, y:rand, z:-1.0}, false, false); 
+            ModelModule::set_joint_translate(fighter.module_accessor, Hash40::new("waist"), &Vector3f{x:3.0, y:0.0, z:rand_waist}, false, false); 
```
random programmatic animation adjustments in opff... I hate it but it might be necessary for what you're doing? Honestly DSpecial charging should work based on two separate animations and statuses like a smash attack (but I won't ask you to make that change)

SuddyN commented on romfs/source/fighter/dedede/motion/body/motion_patch.yaml:
```diff
@@ -117,7 +123,7 @@ special_air_s_get:
     cancel_frame: 53
 special_hi_failure:
   extra:
-    cancel_frame: 41
+    cancel_frame: 98
```
is this to delay freefall?

SirTanknSpank committed:
Requested changes + Side B/Recatch refactor

- Reformatting motion rates
- Adding side b status scripts
- Refactoring gordo recatch
- Also fixes some small bugs with side b and recatch
